scipy/special/_cosine.c

```diff
@@ -19,6 +19,9 @@
 
 #include "cephes/polevl.h"
 
+#if !defined(M_PI)
+#define M_PI 3.1415926535897932384626433832795028841971
```

Comment:
Can you try this?
```suggestion
#define M_PI 0x1.921fb54442d18p+1
```
That should be the exact 64 bit floating point value that is closest to π.   I don't know if this will make a difference, but it is worth a shot.  Without having the actual compiler to test, I don't know if it parses the extra long decimal representation in a way that ends up being different from the best 64 bit approximation to π.

Reply:
> Not only should you not rely on M_PI being available, ...

And I even knew that, but forgot to handle it here.

> ... but if it is available, the exact numerical value is going to be a compiler-dependent approximation to \pi. It's very hard to believe that all implementations are going to be off by about `np.finfo(np.float64).eps - 1e-16`.

The code was written under the assumption that a C `double` would be a IEEE 754 64 bit floating point value.   The C language standard does not make such a guarantee, and if a C `double` is something else, then you are correct, this code won't work as I expected.

Reply:
Or just see what the actual defines in `math.h` are. Just checked in GCC 9.3.0 on macOS, and it's
```
#define M_PI        3.14159265358979323846264338327950288   /* pi             */
```

Reply:
Giving this a go. That's a funky-looking way to write a floating-point number.